src/frontend/src/scheduler/hummock_snapshot_manager.rs

```diff
@@ -13,125 +13,215 @@
 }
 
-#[derive(Default)]
 struct HummockSnapshotManagerCore {
-    is_outdated: bool,
-    last_pinned: u64,
     /// Record the query ids that pin each snapshot.
     /// Send an `unpin_snapshot` RPC when a snapshot is not pinned any more.
     epoch_to_query_ids: BTreeMap<u64, HashSet<QueryId>>,
+    meta_client: Arc<dyn FrontendMetaClient>,
+    last_unpin_snapshot: Arc<AtomicU64>,
 }
 
 impl HummockSnapshotManagerCore {
-    fn new() -> Self {
+    fn new(meta_client: Arc<dyn FrontendMetaClient>) -> Self {
         Self {
             // Initialize by setting `is_outdated` to `true`.
-            is_outdated: true,
-            ..Default::default()
+            meta_client,
+            epoch_to_query_ids: BTreeMap::default(),
+            last_unpin_snapshot: Arc::new(AtomicU64::new(0)),
+        }
+    }
+
+    async fn get_epoch_for_query(
+        &mut self,
+        batches: &mut Vec<(QueryId, Callback<SchedulerResult<u64>>)>,
+    ) -> u64 {
+        let ret = self.meta_client.get_epoch().await;
+        match ret {
+            Ok(epoch) => {
+                let queries = match self.epoch_to_query_ids.get_mut(&epoch) {
+                    None => {
+                        self.epoch_to_query_ids.insert(epoch, HashSet::default());
+                        self.epoch_to_query_ids.get_mut(&epoch).unwrap()
+                    }
+                    Some(queries) => queries,
+                };
+                for (id, cb) in batches.drain(..) {
+                    queries.insert(id);
+                    let _ = cb.send(Ok(epoch));
+                }
+                epoch
+            }
+            Err(e) => {
+                for (id, cb) in batches.drain(..) {
+                    let _ = cb.send(Err(SchedulerError::Internal(anyhow!(
+                        "Failed to get epoch for query: {:?} because of RPC Error: {:?}",
+                        id,
+                        e
+                    ))));
+                }
+                0
+            }
+        }
+    }
+
+    pub fn release_epoch(&mut self, queries: &mut Vec<(QueryId, u64)>) {
+        for (query_id, epoch) in queries.drain(..) {
+            let query_ids = self.epoch_to_query_ids.get_mut(&epoch);
+            if let Some(query_ids) = query_ids {
+                query_ids.remove(&query_id);
+                if query_ids.is_empty() {
+                    self.epoch_to_query_ids.remove(&epoch);
+                }
+            }
         }
     }
+
+    pub fn unpin_snapshot_before(&mut self, last_committed_epoch: u64) {
+        // Check the min epoch which still exists running query. If there is no running query,
+        // we shall unpin snapshot with the last committed epoch.
+        let min_epoch = match self.epoch_to_query_ids.first_key_value() {
+            Some((epoch, _)) => *epoch,
+            None => last_committed_epoch,
+        };
+
+        if min_epoch <= self.last_unpin_snapshot.load(Ordering::Acquire) {
+            return;
+        }
+
```

Comment:
I'm confused why we needn't ensure no running query in min_epoch.

Reply:
I do not understand what you mean. 
All running query will register their QueryId in `epoch_to_query_ids` just after they started. And they will unregister themself after query end so that we can calculate it by `BTreeMap`.